go.mod

```diff
@@ -4,6 +4,6 @@ go 1.15
 
 require (
 	github.com/benoitkugler/textlayout v0.0.5
-	golang.org/x/image v0.0.0-20210607152325-775e3b0c77b9
-	golang.org/x/text v0.3.7 // indirect
+	github.com/gioui/uax v0.2.1-0.20220819135011-cda973fac06d
```

Comment:
It makes me nervous that a text project aiming to be toolkit agnostic would import repos from one of the toolkit accounts.
If it needs to be forked for better maintenance would this (`go-text`) not be the right place to do it?

Reply:
Gio forked and improved the code because Gio actively depends on it in our main branch. Since these changes were in limbo at the time, it wasn't clear to me that I should put it in go-text. That being said, I've already offered to move it to this org in the top level repo comments. There's nothing territorial here. It just made sense at the time to put it in Gio's org.

Reply:
Absolutely - I did not mean to imply it was territorial, just that it should be re-considered for the purpose of this PR.

Reply:
If we decide that the content of the dependency is okay, I'll move it into go-text before we merge this.

Reply:
I'd be interested to hear more from @benoitkugler - is their replacement a better alternative?

Reply:
I had a lot of free time last week, and I was able to work on the replacement implementation I was referring to.

To give more context and answer @whereswaldon question, I think we should use this over `github.com/gioui/uax` for two reasons : 

- The code in `uax` is very high level : the implementation strategy is very refined, but it makes the code quite complicated to understand. I'm not referring to the Unicode rules, some of them being inherently intricate, but rather to the general logic flow : you need to understand several concepts and structures (such as `NfaStateFn`, `Recognizer`, `UnicodePublisher`, `RuneSubscriber`, ...) to be able to understand the code. 
- The abstraction has a performance cost : even with the nice work on the `gioui` fork, both speed and memory allocations are significantly higher than a "simpler", more linear, approach

First precision : the performance argument is probably not very strong for now, but still I think that a low level, shared, text framework should aim at good performances.
Second precision : I actually didn't understand the logic in `uax`, which makes my first point a bit biased.. but I suspect I won't be the only one..

I have pushed two branches : 

- [segmenter](https://github.com/go-text/typesetting/tree/segmenter/segmenter) is a playground with both implementations available, so that you can easily run the benchmarks in the `segmenter` package, and check that my performance point is accurate
- [line-wrapping-segmenter](https://github.com/go-text/typesetting/tree/line-wrapping-segmenter/segmenter) uses  this new package as a replacement for `uax` with minimal changes (and tests passing).

The Unicode conformance tests for [Graphemes](https://www.unicode.org/Public/15.0.0/ucd/auxiliary/GraphemeBreakTest.txt) and [Line Breaks](https://www.unicode.org/Public/15.0.0/ucd/auxiliary/LineBreakTest.txt) are passing.

Regarding the license, I have borrowed the iterator pattern from `uax` and the concept of the break attributes from the Pango source code, but I've written the go code from the ground up, and thus the license should not be an issue here. 

I propose this package as an alternative, but I will let you decide what is best. 

(If using this new `segmenter` package sounds good to you, we should probably start by reviewing its code before  merging this PR.)

Reply:
I'm looking at your implementation now @benoitkugler and it's very impressive. I think there's little doubt that we should switch to it. I do think I'll have some questions/suggestions about the API though. I just need some more time to look and experiment.

Reply:
> I do think I'll have some questions/suggestions about the API though.

Sure thing, I'll be happy to include your suggestions !